diff -ur GnomeSensors-0.2.0/src/chips.c GnomeSensors-0.2.0-via/src/chips.c
--- GnomeSensors-0.2.0/src/chips.c	Wed Nov 24 21:27:41 1999
+++ GnomeSensors-0.2.0-via/src/chips.c	Mon Oct 14 23:38:57 2002
@@ -223,6 +223,16 @@
 }
 
 /* ==========================================================================
+ * read_via686a
+ * =========================================================================*/
+void read_via686a(machine *all_sensors, const sensors_chip_name *name)
+{
+	get_sensors( all_sensors, SENSORS_VIA686A_TEMP, SENSORS_VIA686A_TEMP3_OVER, name, CELSIUS);	
+	get_sensors( all_sensors, SENSORS_VIA686A_FAN1, SENSORS_VIA686A_FAN2, name, RPM);	
+	get_sensors( all_sensors, SENSORS_VIA686A_IN0, SENSORS_VIA686A_IN4, name, VOLTS);	
+}
+
+/* ==========================================================================
  * fetch_sensor_for_display
  * =========================================================================*/
 void fetch_sensor_for_display( machine *all_sensors,int index, char *name, char *value, char *units)
@@ -276,6 +286,8 @@
 		*chip = GL518;
 	else if (!strcmp(name, "gl520sm"))
 		*chip = GL520;
+	else if (!strcmp(name, "via686a"))
+		*chip = VIA686A;
 	else if ((!strcmp(name,"w83781d")) ||
 			(!strcmp(name,"w83782d")) ||
 			(!strcmp(name,"w83783s")) ||
@@ -350,7 +362,10 @@
 				break;
 			case LM75:
 				read_lm75(all_sensors, &name);
-				
+				break;
+			case VIA686A:
+				read_via686a(all_sensors, &name);
+
 			default:
 				break;
 		}
diff -ur GnomeSensors-0.2.0/src/chips.h GnomeSensors-0.2.0-via/src/chips.h
--- GnomeSensors-0.2.0/src/chips.h	Sat Nov 20 11:12:17 1999
+++ GnomeSensors-0.2.0-via/src/chips.h	Mon Oct 14 23:32:37 2002
@@ -48,7 +48,8 @@
 		GL518,
 		GL520,
 		W83781D,    /* w83781d, w83782d, w83783s, w83627hf*/
-		MAXILIFE    /* maxilife-cg, maxilife-co, maxilife-as */
+		MAXILIFE,   /* maxilife-cg, maxilife-co, maxilife-as */
+		VIA686A
 	} known_chips;
 	
 	typedef enum _units
